datafusion/src/physical_plan/hash_join.rs

```diff
@@ -751,12 +788,19 @@ fn build_join_indexes(
 }
 
 macro_rules! equal_rows_elem {
-    ($array_type:ident, $l: ident, $r: ident, $left: ident, $right: ident) => {{
+    ($array_type:ident, $l: ident, $r: ident, $left: ident, $right: ident, $null_equal_safe: ident) => {{
         let left_array = $l.as_any().downcast_ref::<$array_type>().unwrap();
         let right_array = $r.as_any().downcast_ref::<$array_type>().unwrap();
 
         match (left_array.is_null($left), right_array.is_null($right)) {
             (false, false) => left_array.value($left) == right_array.value($right),
+            (true, true) => {
```

Comment:
I am concerned bout the performance impact.
It would expect it to be optimized away, as it's using a macro for this and I would expect the branch to be removed based on the boolean constant, but not sure about that.

Reply:
Also the `is_null` check and downcasting per item, I would expect this has a potential higher impact, even if the added code has some non-zero cost.